reactor-netty-http-brave/src/test/java/reactor/netty/http/brave/ITTracingHttpClientDecoratorTest.java

```diff
@@ -89,50 +84,37 @@ protected void post(HttpClient client, String pathIncludingQuery, String body) {
 	}
 
 	@Test
+	@Ignore("TODO: fix broken context")
 	public void currentSpanVisibleToUserHandler() {
-		AtomicReference<HttpHeaders> headers = new AtomicReference<>();
-		DisposableServer disposableServer = null;
-		TraceContext parent = newTraceContext(SamplingFlags.SAMPLED);
-		try (CurrentTraceContext.Scope scope = currentTraceContext.newScope(parent)) {
-			disposableServer =
-					HttpServer.create()
-					          .port(0)
-					          .handle((req, res) -> {
-					              headers.set(req.requestHeaders());
-					              return res.sendString(Mono.just("test"));
-					          })
-					          .bindNow();
-
-			client.port(disposableServer.port())
-			      .request(HttpMethod.GET)
-			      .uri("/")
-			      .send((req, out) -> {
-			          TraceContext traceContext = req.currentContextView().getOrDefault(TraceContext.class, null);
-			          if (traceContext != null) {
-			              req.header("test-id", traceContext.traceIdString());
-			          }
-			          return out;
-			      })
-			      .responseContent()
-			      .aggregate()
-			      .block(Duration.ofSeconds(30));
-
-			assertThat(headers.get()).isNotNull();
-			assertThat(headers.get().get("x-b3-traceId")).isEqualTo(headers.get().get("test-id"));
-		}
-		finally {
-			if (disposableServer != null) {
-				disposableServer.disposeNow();
-			}
+		server.enqueue(new MockResponse());
+
+		TraceContext invocationContext = newTraceContext(SamplingFlags.SAMPLED);
+		try (Scope ws = currentTraceContext.newScope(invocationContext)) {
+			client.request(HttpMethod.GET)
+					.uri("/")
+					.send((req, out) -> {
+						// currentTraceContext should either be the invocationContext or the
+						// client span context depending on if the request has been sent or not.
+						req.header("my-id", currentTraceContext.get().traceIdString());
```

Comment:
this will NPE

all client instrumentation either instrument their callbacks or rely on a context integration that does the same. For example, armeria has a context integration which allows all callbacks to see the current span.

Reply:
At this point

https://github.com/reactor/reactor-netty/blob/631f7a67269ac450195d2336714edb9098e6736f/reactor-netty-http-brave/src/main/java/reactor/netty/http/brave/TracingHttpClientDecorator.java#L304-L310

Reactor Netty has only the invocation context

Later when a request is received  we can take the span context

https://github.com/reactor/reactor-netty/blob/631f7a67269ac450195d2336714edb9098e6736f/reactor-netty-http-brave/src/main/java/reactor/netty/http/brave/TracingHttpClientDecorator.java#L198

We can make this AtomicReference as SpanCutomizer and later on if it was NULL to put span context.

Wdyt?

Reply:
I think we are mixing contexts. CurrentTraceContext is what will allow transparent continuation of spans. it seems you are adding data into reactor context. Unless they are integrated, then the trace will still break unless someone manually looks at reactor context, correct?

Reply:
yes

Reply:
so we can add a different test that looks similar to test the contents of reactor context. maybe that one can pass?

Reply:
yes something similar to what we have for the server?
https://github.com/reactor/reactor-netty/blob/631f7a67269ac450195d2336714edb9098e6736f/reactor-netty-http-brave/src/test/java/reactor/netty/http/brave/ITTracingHttpServerDecoratorTest.java#L66

Reply:
haha yes I didn't notice that the server is cheating :D the normal "/child" path should use implicit (currentTraceContext) instead of explicit. that way people can know what's supported and what isn't.